splitgraph/ingestion/sqlite/__init__.py

```diff
@@ -153,6 +181,36 @@ def get_preview_rows(
     ]
 
 
+SQLITE_IMPLICIT_ROWID_COLUMN_NAME = "ROWID"
+# copy 1000 rows in a single iteration
+DEFAULT_BATCH_SIZE = 1000
+
+
+def get_select_query(
+    table_name: str,
+    primary_keys: List[str],
+    end_of_last_batch: Optional[sqlite3.Row],
+    batch_size: int,
+):
+    effective_pks = primary_keys if len(primary_keys) > 0 else [SQLITE_IMPLICIT_ROWID_COLUMN_NAME]
+    pk_column_list = ", ".join([_quote_ident(col) for col in effective_pks])
+    where_clause = "true"
+    if end_of_last_batch is not None:
+        last_batch_end_tuple = ", ".join(
+            [emit_value(end_of_last_batch[col]) for col in effective_pks]
+        )
+        where_clause = f"({pk_column_list}) > ({last_batch_end_tuple})"
+    return "SELECT {}* FROM {} WHERE {} ORDER BY {} ASC LIMIT {}".format(  #  nosec
+        # add the implicit rowid column to the select if no explicit primary
+        # key columns exist on table, based on: https://www.sqlite.org/withoutrowid.html
+        f"{SQLITE_IMPLICIT_ROWID_COLUMN_NAME}, " if len(primary_keys) == 0 else "",
+        _quote_ident(table_name),
+        where_clause,
+        pk_column_list,
+        batch_size,
+    )  #  nosec
```

Comment:
You can avoid trying to reproduce SQLite's identifier quoting by emitting a string with placeholders and returning both the string and the arguments to interpolate, so instead of:

```
"SELECT ..... WHERE (col1, col2, col3) > ('some-val', 42, 'some-other-val)"
```

you'd return

```
"SELECT ..... WHERE (col1, col2, col3) > (%s, %s, %s)"
("some-val", 42, "some-other-val")
```

and call it like:

```python
query, args = get_select_query(...)
table_contents = query_connection(con, query, args)
```

Reply:
I thought about that, but the problem is that the first batch has the condition `WHERE TRUE`, so the number of arguments depends on which batch is being queried, which seemed more complex.

Reply:
In that case, your function can return `"WHERE TRUE...", ()` (empty tuple) or even `"WHERE %s", (True)` and the caller can execute the statement in the same way. IMO it's less complex and more robust than making sure all bases are covered with manual quoting.